File: Algoexpert/Sum of Linked Lists/cpp/sum_of_two_linked_lists.cpp
#include <bits/stdc++.h>

using namespace std;

class LinkedList {
public:
    int value;
    LinkedList *next = nullptr;

    LinkedList(int value) { this->value = value; }
};

LinkedList *addMany(LinkedList *linkedList, vector<int> values) {
    auto current = linkedList;
    while (current->next != nullptr) {
        current = current->next;
    }
    for (int value : values) {
        current->next = new LinkedList(value);
        current = current->next;
    }
    return linkedList;
}

vector<int> getNodesInArray(LinkedList *linkedList) {
    vector<int> nodes;
    auto current = linkedList;
    while (current != nullptr) {
        nodes.push_back(current->value);
        current = current->next;
    }
    return nodes;
}

class Solution {
    public:
        LinkedList *sumOfLinkedLists(LinkedList *linkedListOne,
                             LinkedList *linkedListTwo) {
            auto newLinkedListHeadPointer = new LinkedList(0);
            auto currentNode = newLinkedListHeadPointer;
            int carry = 0;

            auto nodeOne = linkedListOne;
            auto nodeTwo = linkedListTwo;

            while(nodeOne != nullptr || nodeTwo != nullptr || carry != 0) {

                int valueOne = nodeOne != nullptr ? nodeOne->value : 0;
                int valueTwo = nodeTwo != nullptr ? nodeTwo->value : 0;
                int sumOfValues = valueOne + valueTwo + carry;

                int newValue = sumOfValues % 10;
                auto newNode = new LinkedList(newValue);
                currentNode->next = newNode;
                currentNode = currentNode->next;

                carry = sumOfValues / 10;
                nodeOne = nodeOne != nullptr ? nodeOne->next : nullptr;
                nodeTwo = nodeTwo != nullptr ? nodeTwo->next : nullptr;
            }

            return newLinkedListHeadPointer->next;
        }

};


int main() {

    Solution solution;
    auto ll1 = addMany(new LinkedList(2), {4, 7, 1});
    auto ll2 = addMany(new LinkedList(9), {4, 5});

    auto result = solution.sumOfLinkedLists(ll1, ll2);

    for(auto res : getNodesInArray(result) ) {
        cout << res << endl;
    }

    return 0;
}

File: Algoexpert/Sum of Linked Lists/python/sum_of_two_linked_lists.py
import unittest

class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None


def sumOfLinkedLists(linkedListOne, linkedListTwo):
    
	newLinkedListHeadPointer = LinkedList(0)
	currentNode = newLinkedListHeadPointer
	carry = 0
	
	nodeOne = linkedListOne
	nodeTwo = linkedListTwo
	
	while nodeOne or nodeTwo or carry != 0:
		
		valueOne = nodeOne.value if nodeOne is not None else 0
		valueTwo = nodeTwo.value if nodeTwo is not None else 0
		sumOfValues = valueOne + valueTwo + carry
		
		newValue = sumOfValues % 10
		newNode = LinkedList(newValue)
		currentNode.next = newNode
		currentNode = currentNode.next
		
		carry = sumOfValues // 10
		nodeOne = nodeOne.next if nodeOne is not None else None
		nodeTwo = nodeTwo.next if nodeTwo is not None else None
		
	return newLinkedListHeadPointer.next

class LinkedList(LinkedList):
    def addMany(self, values):
        current = self
        while current.next is not None:
            current = current.next
        for value in values:
            current.next = LinkedList(value)
            current = current.next
        return self


def getNodesInArray(output):
    nodes = []
    current = output
    while current is not None:
        nodes.append(current.value)
        current = current.next
    return nodes


class TestProgram(unittest.TestCase):
    def test_case_1(self):
        ll1 = LinkedList(2).addMany([4, 7, 1])
        ll2 = LinkedList(9).addMany([4, 5])
        expected = LinkedList(1).addMany([9, 2, 2])
        actual = sumOfLinkedLists(ll1, ll2)
        self.assertEqual(getNodesInArray(actual), getNodesInArray(expected))
        
if __name__ == "__main__":
    unittest.main()